File: backend/backend-java/src/main/java/com/cayena/backend/controllers/ProductController.java
package com.cayena.backend.controllers;

import com.cayena.backend.dtos.requesties.ProductRequest;
import com.cayena.backend.dtos.requesties.QuantityProductRequest;
import com.cayena.backend.dtos.responses.ProductAllResponse;
import com.cayena.backend.dtos.responses.ProductResponse;
import com.cayena.backend.services.impl.ProductService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort.Direction;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;

@RestController
@RequestMapping("/api/v1/products")
public class ProductController {

    @Autowired
    private ProductService service;

    @PostMapping
    public ResponseEntity<ProductResponse> saveProduct(@RequestBody @Valid ProductRequest request) {
        ProductResponse response = service.save(request);
        return new ResponseEntity<>(response, HttpStatus.CREATED);
    }

    @GetMapping
    public ResponseEntity<Page<ProductResponse>> getAllProducts(

            @RequestParam(value = "page", defaultValue = "0") Integer page,
            @RequestParam(value = "linesPerPage", defaultValue = "10") Integer linesPerPage,
            @RequestParam(value = "direction", defaultValue = "ASC") String direction,
            @RequestParam(value = "orderBy", defaultValue = "name") String orderBy
    ){
        PageRequest pageRequest = PageRequest.of(page, linesPerPage, Direction.valueOf(direction), orderBy );
        Page<ProductResponse> list = service.getAllProducts(pageRequest);
        return ResponseEntity.ok().body(list);
    }

    @GetMapping("/{id}")
    public ResponseEntity<ProductAllResponse> getProductById(@PathVariable Long id){
        ProductAllResponse response = service.getProductById(id);
        return ResponseEntity.ok().body(response);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ProductResponse> deleteProduct(@PathVariable Long id){
        service.delete(id);
        return ResponseEntity.noContent().build();
    }
    @PutMapping("/{id}")
    public ResponseEntity<ProductResponse> updateProduct(@PathVariable Long id, @RequestBody @Valid ProductRequest request){
        ProductResponse response = service.update(id, request);
        return ResponseEntity.ok().body(response);
    }
    @PatchMapping("/{id}")
    public ResponseEntity<ProductResponse> increasOrDecreaseQuantityStock(@PathVariable Long id, @RequestBody @Valid QuantityProductRequest request){
        ProductResponse response = service.updateQuantityStock(id, request);
        return ResponseEntity.ok().body(response);
    }
}

File: backend/backend-kotlin/src/main/kotlin/com/cayena/backendkotlin/controllers/ProductController.kt
package com.cayena.backendkotlin.controllers

import com.cayena.backendkotlin.dtos.requests.ProductRequest
import com.cayena.backendkotlin.dtos.requests.QuantityProductRequest
import com.cayena.backendkotlin.dtos.responses.ProductResponse
import com.cayena.backendkotlin.services.impl.ProductService
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.web.PageableDefault
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.util.UriComponentsBuilder
import javax.validation.Valid

@RestController
@RequestMapping("api/v1/products")
class ProductController(
    val service: ProductService
) {

    @PostMapping
    fun createProduct(
        @Valid @RequestBody productRequest: ProductRequest,
        uriComponentsBuilder: UriComponentsBuilder
    ): ResponseEntity<ProductResponse> {

        val productResponse = service.createProduct(productRequest)
        val uri = uriComponentsBuilder.path("id").build().toUri()
        return ResponseEntity.created(uri).body(productResponse)
    }

    @GetMapping("/{id}")
    fun getProductById(
        @PathVariable id: Long
    ): ResponseEntity<ProductResponse> {
        return ResponseEntity.ok().body(service.getProductById(id))
    }

    @GetMapping
    fun getAllProducts(
        @RequestParam(required = false) nameProduct: String?,
        @PageableDefault(size = 10, sort = ["productId"], direction = Sort.Direction.DESC) pageable: Pageable
    ): ResponseEntity<Page<ProductResponse>> {
        val products: Page<ProductResponse> = service.getAllProducts(nameProduct, pageable)
        return ResponseEntity.ok().body(products)
    }

    @DeleteMapping("/{id}")
    fun deleteProduct(@PathVariable id: Long): ResponseEntity<Void> {
        service.deleteProduct(id)
        return ResponseEntity.noContent().build()
    }

    @PutMapping("/{id}")
    fun updateProduct(
        @PathVariable id: Long,
        @Valid @RequestBody productRequest: ProductRequest
    ): ResponseEntity<ProductResponse> {
        val product: ProductResponse = service.updateProduct(id, productRequest)
        return ResponseEntity.ok().body(product)
    }

    @PatchMapping("/{id}")
    fun updateQuantityStockProduct(
        @PathVariable id: Long,
        @RequestBody @Valid quantityProductRequest: QuantityProductRequest
        ): ResponseEntity<ProductResponse>{
        val product: ProductResponse = service.updateQuantityStock(id, quantityProductRequest)
        return ResponseEntity.ok().body(product)
    }
}